Sources/MapboxNavigation/NavigationMapView+BuildingHighlighting.swift

```diff
@@ -23,17 +23,20 @@ extension NavigationMapView {
                                    in3D extrudesBuildings: Bool = true,
                                    extrudeAll: Bool = false,
                                    completion: ((_ foundAllBuildings: Bool) -> Void)? = nil) {
-        var foundBuildingIds = Set<Int64>()
+        highlightBuildings = highlightBuildings.filter{ coordinates.contains($0.key) }
```

Comment:
`highlightBuildings` effectively caches building identifiers between calls to this method, but this method doesn’t consult the cache before querying and adding results to it. Do we really need the cache?

Reply:
I think that we should not use intermediate cache in `NavigationMapView` specifically. Not sure if it's the right word to describe it, but they should be *idempotent*. 

IMHO all logic related to building highlighting during navigation specifically is meant to be implemented by users of this API (in this case it's `NavigationViewController` or `CarPlayNavigationViewController`). Both of them conform to `BuildingHighlighting` with default implementation in `attemptToHighlightBuildings(_:,navigationMapView:completion:)`.

Reply:
I think we need this cache. For example, right now on the main branch. If we add a waypoint and highlight it, pan the mapView and move the waypoint out of bounds, and then choose another waypoint. Because the previous waypoint coordinate is out of mapView bound, the query will fail. So we could only highlight the last waypoint in the mapView bounds. It's not an expected behavior.
With the cache, when we choose multiple waypoint, if the previous one has been highlighted before, its identifier will be kept and highlighted even it's out of the map bounds.

Reply:
I think I understand now how `highlightedBuildingIdentifiersByCoordinate` prevents the problem you’re describing, but it’s kind of confusing. The method never checks whether a coordinate is already in `highlightedBuildingIdentifiersByCoordinate` before it queries rendered features at the coordinate. Instead, it relies on the fact that the query will come back with no results, allowing a fallback to the existing value in `highlightedBuildingIdentifiersByCoordinate`.

So `highlightedBuildingIdentifiersByCoordinate` isn’t actually a cache in the sense of avoiding redundant work. Rather, it just memoizes previous successful results. Memoization normally doesn’t require a dictionary, but it does in this case because we need to keep track of both coordinates and building IDs. The initial filter mostly addresses the question of clearing the dictionary. Because of the filter, every call to this method effectively shrinks `highlightedBuildingIdentifiersByCoordinate` to just the results for the passed-in coordinates. (It’s possible that the method could be called with slightly different coordinates for the same building, which would effectively clear the dictionary.)